python_modules/dagster/dagster/_core/definitions/metadata/__init__.py

```diff
@@ -1289,7 +1291,26 @@ class MyMetadataEntries(NamedspacedMetadataEntries):
             if len(splits) == 2:
                 namespace, key = splits
                 if namespace == cls.namespace() and key in cls.__fields__:
-                    kwargs[key] = value.value if isinstance(value, MetadataValue) else value
+                    if isinstance(value, MetadataValue):
+                        # Based on type annotation, potentially coerce the metadata value to its
+                        # inner value.
+                        # E.g. if the annotation is Optional[float] and the value is
+                        # FloatMetadataValue, construct the MetadataSet using the inner float.
+                        annotation = cls.__fields__[key].annotation
+                        accepted_types = (
```

Comment:
Might be nice add explicit types for this, since these are showing up as `Any` in local pyright.

Reply:
👍 added